safe-logging/src/main/java/com/palantir/logsafe/SafeArg.java

```diff
@@ -16,15 +16,21 @@
 
 package com.palantir.logsafe;
 
+import java.util.function.Supplier;
+
 /** A wrapper around an argument known to be safe for logging. */
 public final class SafeArg<T> extends Arg<T> {
 
-    private SafeArg(String name, T value) {
-        super(name, value);
+    private SafeArg(String name, Supplier<T> lazyValue) {
+        super(name, lazyValue);
     }
 
     public static <T> SafeArg<T> of(String name, T value) {
-        return new SafeArg<>(name, value);
+        return new SafeArg<>(name, () -> value);
```

Comment:
Would highly recommend adding some JMH benchmarks to compare before and after here as this is going to add more allocations on potentially hot paths. I have a feeling we probably want a lazy and static impl for each safe and unsafe, and you can move most code into an abstract base class